In [3]:
import psycopg2
import pandas as pd

# Устанавливаем соединение с базой данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Параметр для ограничения количества записей
n = 10

# SQL-запрос для выборки данных:
# Из таблицы public.vacancies выбираем все столбцы
# Ограничиваем выборку указанным числом записей (LIMIT n)
query = f"""
    SELECT *
    FROM public.vacancies
    LIMIT {n}
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный датафрейм
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1799951673.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   
5  55906072                                          Корректор   
6  55523207                   React Front-end Middle Developer   
7  55050261                                Инженер-программист   
8  55583432                            Системный администратор   
9  55312414        Ведущий специалист группы разработки систем   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   
5  Корректура текстов\tГрамотность\tРабота в кома...       Полный день   
6       JavaScript\tTypeScript\tCSS3\tNode.js\tHTML5     Гибкий график   
7  Java SE\tSpring Framework\tSpring Boot\tGit\tS...       Полный день   
8  Linux\tАдминистрирование сетевого оборудования...       Полный день   
9  1С программирование\t1С: Предприятие 8\t1С: До...       Полный день   

           experience           employment  salary_from  salary_to  area_id  \
0           Нет опыта     Полная занятость      64000.0        NaN     1516   
1       От 3 до 6 лет     Полная занятость          NaN        NaN      130   
2       От 3 до 6 лет     Полная занятость          NaN        NaN      160   
3       От 3 до 6 лет     Полная занятость          NaN        NaN      159   
4           Нет опыта  Частичная занятость      15000.0        NaN     1955   
5           Нет опыта     Полная занятость          NaN        NaN     2323   
6  От 1 года до 3 лет     Полная занятость          NaN        NaN     2759   
7           Нет опыта     Полная занятость      40000.0    60000.0       49   
8  От 1 года до 3 лет     Полная занятость      50000.0        NaN        3   
9       От 3 до 6 лет     Полная занятость          NaN        NaN       47   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740  
5       584934  
6       204511  
7      4448636  
8        55126  
9      1413754

In [4]:
import psycopg2
import pandas as pd

# Устанавливаем соединение с базой данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Параметр для ограничения количества записей
n = 10

# SQL-запрос для выборки данных:
# Из таблицы public.areas выбираем все столбцы
# Ограничиваем выборку указанным числом записей (LIMIT n)
query = f"""
    SELECT *
    FROM public.areas
    LIMIT {n}
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный датафрейм
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\799055341.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                           name
0  2758                        Тбилиси
1     8                         Майкоп
2  1180                       Нерюнгри
3  1240                    Новокузнецк
4     2                Санкт-Петербург
5     1                         Москва
6  2061  Подольск (Московская область)
7  1545                   Новочеркасск
8  2759                        Ташкент
9  3695                   Промышленная

In [5]:
import psycopg2
import pandas as pd

# Устанавливаем соединение с базой данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Параметр для ограничения количества записей
n = 10

# SQL-запрос для выборки данных:
# Из таблицы public.employers выбираем все столбцы
# Ограничиваем выборку указанным числом записей (LIMIT n)
query = f"""
    SELECT *
    FROM public.employers
    LIMIT {n}
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный датафрейм
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1337400288.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  area
0     2393                   Программный Продукт, ИТ-компания     1
1    72977                                          БАРС Груп    88
2     3155  Бест, Торгово-производственная компания, Екате...     3
3      675                                   КОРУС Консалтинг     2
4  1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88
5   213397                                            МойОфис     1
6    41862                                             Контур     3
7  1080425  Челябинский государственный университет, ФГБОУ...   104
8     4080                           Axenix (ранее Accenture)     1
9     3415                                       ИНФОРМЗАЩИТА     1

In [6]:
import psycopg2
import pandas as pd

# Устанавливаем соединение с базой данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Параметр для ограничения количества записей
n = 10

# SQL-запрос для выборки данных:
# Из таблицы public.industries выбираем все столбцы
# Ограничиваем выборку указанным числом записей (LIMIT n)
query = f"""
    SELECT *
    FROM public.industries
    LIMIT {n}
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный датафрейм
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\311349466.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name
0   7.540                Разработка программного обеспечения
1   7.539  Системная интеграция,  автоматизации технологи...
2  27.550              Безалкогольные напитки (производство)
3  27.551  Безалкогольные напитки (продвижение, оптовая т...
4  13.664             Управление и эксплуатация недвижимости
5   5.466                                   Складские услуги
6  39.442                             Вуз, ссуз колледж, ПТУ
7  44.393                              Консалтинговые услуги
8   5.467                      Курьерская, почтовая доставка
9  41.525                                   Интернет-магазин

In [7]:
import psycopg2
import pandas as pd

# Устанавливаем соединение с базой данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Параметр для ограничения количества записей
n = 10

# SQL-запрос для выборки данных:
# Из таблицы public.employers_industries выбираем все столбцы
# Ограничиваем выборку указанным числом записей (LIMIT n)
query = f"""
    SELECT *
    FROM public.employers_industries
    LIMIT {n}
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный датафрейм
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\3965408241.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_id industry_id
0         2393       7.540
1         2393       7.539
2        72977       7.540
3        72977       7.539
4         3155      27.550
5         3155      27.551
6         3155      13.664
7         3155       5.466
8          675       7.540
9       213397       7.540

In [8]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.vacancies
query = """
    SELECT
        COUNT(DISTINCT id)
    FROM public.vacancies
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1082563796.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  49197

In [9]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.employers
query = """
    SELECT
        COUNT(DISTINCT id)
    FROM public.employers
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\869082355.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  23501

In [10]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.areas
query = f"""
    SELECT
        COUNT(DISTINCT id)
    FROM public.areas
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\2962645852.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0   1362

In [11]:
# SQL-запрос для подсчета количества уникальных значений:
# Подсчитываем количество уникальных значений в столбце id из таблицы public.industries
query = f"""
    SELECT
        COUNT(DISTINCT id)
    FROM public.industries
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\733051974.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0    294

In [12]:
# SQL-запрос для подсчета количества вакансий по регионам:
# Из таблицы public.vacancies объединяем данные с таблицей public.areas по полю area_id
# Считаем количество вакансий для каждого региона, сортируем по убыванию количества и берем топ-5
query = f"""
    SELECT 
        a.name,              -- Название региона
        COUNT(v.id) AS count -- Количество вакансий в регионе
    FROM public.vacancies v
    JOIN public.areas a
        ON v.area_id = a.id  -- Соединение таблиц по area_id
    GROUP BY a.name          -- Группировка по названию региона
    ORDER BY count DESC      -- Сортировка по количеству вакансий в убывающем порядке
    LIMIT 5                  -- Ограничение на 5 записей
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\3773073412.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

### 📊 Анализ результатов запроса: количество вакансий по регионам

#### **1. Распределение вакансий по регионам**
- **Москва** — 5333 вакансии, что делает её безусловным лидером.
- **Санкт-Петербург** — 2851 вакансия, занимает второе место.
- **Минск** — 2112 вакансий, третье место.
- **Новосибирск** — 2006 вакансий, четвёртое место.
- **Алматы** — 1892 вакансии, пятое место.

#### **2. Выводы**
1. **Централизация рынка труда:**
   - Основная концентрация вакансий приходится на Москву и Санкт-Петербург. Эти города являются основными экономическими и деловыми центрами России.

2. **Роль региональных центров:**
   - Минск и Алматы входят в топ-5, что связано с международными и региональными компаниями, развивающими бизнес в этих городах.
   - Новосибирск, как крупный научно-образовательный центр, показывает стабильный спрос на специалистов.

3. **Дисбаланс регионов:**
   - Разрыв между Москвой и другими городами указывает на высокую степень централизации рабочих мест и компаний.

#### **3. Тенденции и прогнозы**
1. **Москва и Санкт-Петербург**:
   - Эти города останутся лидерами по количеству вакансий, учитывая их экономическое и деловое доминирование.
2. **Минск и Алматы**:
   - Возможен рост вакансий в этих городах благодаря укреплению международного сотрудничества и развитию IT-индустрии.
3. **Региональные центры**:
   - Новосибирск и другие города-миллионники демонстрируют потенциал для дальнейшего роста рынка труда.

#### **4. Рекомендации для дальнейшего анализа**
1. **Категории вакансий**:
   - Определить, какие отрасли преобладают в каждом из городов (например, IT, медицина, финансы).
2. **Зарплаты**:
   - Проанализировать средние зарплаты по регионам для оценки уровня оплаты труда.
3. **Динамика**:
   - Сравнить данные с предыдущими периодами, чтобы выявить тренды изменения количества вакансий.
4. **Вакансии Data Scientist**:
   - Провести отдельный анализ вакансий для специалистов в области анализа данных.

In [13]:
# SQL-запрос для подсчета количества вакансий, где указана зарплата:
# Проверяем, что хотя бы одно из полей salary_from или salary_to не равно NULL
query = f"""
    SELECT
        COUNT(*)
    FROM public.vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\874898699.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  24073

In [14]:
# SQL-запрос для подсчета количества вакансий по графику работы и занятости:
# Группируем данные по полям schedule и employment, сортируем по убыванию количества и берем топ-2
query = f"""
    SELECT
        schedule,           -- График работы (например, полный день, сменный)
        employment,         -- Тип занятости (например, полная занятость, частичная)
        COUNT(*) AS count   -- Количество вакансий для каждой комбинации
    FROM public.vacancies
    GROUP BY schedule, employment -- Группировка по графику и типу занятости
    ORDER BY count DESC           -- Сортировка по количеству вакансий в убывающем порядке
    LIMIT 2                       -- Ограничение на 2 записи
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\3697204009.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


schedule        employment  count
0       Полный день  Полная занятость  35367
1  Удаленная работа  Полная занятость   7802

### 📊 Анализ результатов запроса: количество вакансий по графику работы и типу занятости

#### **1. Результаты**
- **Полный день, Полная занятость** — 35,367 вакансий.
- **Удалённая работа, Полная занятость** — 7,802 вакансии.

#### **2. Выводы**
1. **Доминирует стандартный график:**
   - Наибольшее количество вакансий связано с полным рабочим днём и полной занятостью. Это указывает на то, что компании по-прежнему ориентированы на традиционный формат работы.

2. **Удалённая работа в тренде:**
   - Значительное количество вакансий (7,802) с удалённым графиком указывает на растущую популярность гибких форматов труда. Это, скорее всего, связано с развитием IT-сектора и последствиями пандемии, которые изменили подход к работе.

3. **Тенденции на рынке труда:**
   - Удалённая работа становится конкурентоспособной альтернативой традиционным графикам, особенно в сферах, где производительность не зависит от физического присутствия сотрудников.

#### **3. Рекомендации для дальнейшего анализа**
1. **Сегментация по отраслям:**
   - Выяснить, в каких отраслях преобладают вакансии с удалённой работой.
   - Сравнить популярность удалённой работы в IT и других секторах.

2. **Динамика изменения форматов работы:**
   - Сравнить текущие данные с историческими, чтобы понять, как меняется доля удалённой работы.

3. **Зарплаты:**
   - Сравнить средние зарплаты для вакансий с удалённой работой и традиционным графиком.

4. **Гибридные форматы:**
   - Провести отдельный анализ гибридных графиков (частичная удалённая работа), если такие данные есть.

In [15]:
# SQL-запрос для подсчета количества вакансий по опыту работы:
# Проверяем, что поле experience не равно NULL, группируем по уровню опыта и сортируем по возрастанию количества
query = f"""
    SELECT
        experience,        -- Уровень опыта (например, "Нет опыта", "От 1 года")
        COUNT(*) AS count  -- Количество вакансий для каждого уровня опыта
    FROM public.vacancies
    WHERE experience IS NOT NULL -- Учитываем только строки, где поле experience заполнено
    GROUP BY experience          -- Группируем по уровню опыта
    ORDER BY count ASC           -- Сортируем по количеству вакансий в порядке возрастания
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1723601062.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

### 📊 Анализ результатов запроса: количество вакансий по опыту работы

#### **1. Результаты**
- **Более 6 лет:** 1337 вакансий.
- **Нет опыта:** 7197 вакансий.
- **От 3 до 6 лет:** 14,511 вакансий.
- **От 1 года до 3 лет:** 26,152 вакансии.

#### **2. Выводы**
1. **Спрос на специалистов с небольшим опытом:**
   - Наибольшее количество вакансий ориентировано на специалистов с опытом от 1 года до 3 лет. Это может указывать на то, что компании ищут сотрудников, которые уже приобрели базовые навыки, но ещё не требуют высокой зарплаты.

2. **Шансы для новичков:**
   - Количество вакансий для тех, у кого "Нет опыта" — 7197. Это подтверждает, что рынок предоставляет возможности для начинающих специалистов, особенно в растущих отраслях, таких как IT.

3. **Опытные специалисты менее востребованы:**
   - Вакансии для специалистов с опытом "Более 6 лет" составляют лишь 1337, что может говорить о следующем:
     - Компании реже открывают позиции для высококвалифицированных сотрудников.
     - Возможно, такие специалисты уже трудоустроены и имеют меньшую мобильность на рынке труда.

4. **Баланс спроса на опытных сотрудников:**
   - Вакансий для специалистов с опытом от 3 до 6 лет (14,511) меньше, чем для сотрудников с опытом от 1 года до 3 лет, но всё ещё достаточно, чтобы считать их важной группой на рынке.

#### **3. Тенденции и прогнозы**
1. **Растущий спрос на начинающих специалистов:**
   - Возможно, это связано с активным развитием новых отраслей (например, IT, маркетинг), где компании готовы обучать сотрудников "с нуля".
2. **Стабильность среднего уровня:**
   - Спрос на специалистов с опытом от 1 года до 6 лет остаётся наиболее высоким и стабильным.
3. **Ограниченный рынок для опытных специалистов:**
   - Компании могут предпочитать развивать своих сотрудников, а не привлекать дорогостоящих специалистов с опытом более 6 лет.

#### **4. Рекомендации для дальнейшего анализа**
1. **Анализ по отраслям:**
   - Разбить вакансии по категориям (например, IT, финансы) для каждого уровня опыта.
2. **Сравнение зарплат:**
   - Исследовать средние зарплаты для каждой категории опыта, чтобы оценить привлекательность вакансий.
3. **Динамика спроса:**
   - Выявить тренды изменения количества вакансий для разных уровней опыта по годам.

In [16]:
# SQL-запрос для подсчета количества вакансий у работодателей:
# Соединяем таблицы public.vacancies и public.employers по идентификатору работодателя
# Подсчитываем количество вакансий для каждого работодателя, сортируем по убыванию и берем топ-5
query = f"""
    SELECT
        e.name AS employer_name, -- Название работодателя
        COUNT(v.id) AS vacancy_count -- Количество вакансий
    FROM public.vacancies v
    JOIN public.employers e
        ON e.id = v.employer_id -- Соединение по идентификатору работодателя
    GROUP BY e.name             -- Группируем по названию работодателя
    ORDER BY vacancy_count DESC -- Сортируем по количеству вакансий в убывающем порядке
    LIMIT 5                     -- Ограничиваем результат до 5 записей
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\4047352868.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_name  vacancy_count
0         Яндекс           1933
1     Ростелеком            491
2       Тинькофф            444
3           СБЕР            428
4  Газпром нефть            331

### 📊 Анализ результатов запроса: количество вакансий у топ-5 работодателей

#### **1. Результаты**
- **Яндекс** — 1,933 вакансии.
- **Ростелеком** — 491 вакансия.
- **Тинькофф** — 444 вакансии.
- **СБЕР** — 428 вакансий.
- **Газпром нефть** — 331 вакансия.

#### **2. Выводы**
1. **Доминирующая позиция Яндекса:**
   - Яндекс лидирует с большим отрывом, предлагая почти в 4 раза больше вакансий, чем Ростелеком, занимающий второе место. Это связано с активным развитием IT-направления и высокими темпами роста компании.

2. **Крупные компании других отраслей:**
   - **Ростелеком** (491) и **СБЕР** (428) активно ищут сотрудников, что связано с их цифровой трансформацией и развитием новых технологий.
   - **Газпром нефть** (331) представляет сектор энергетики, где спрос на специалистов остаётся стабильным.

3. **Тренд на цифровизацию:**
   - Большинство лидеров (Яндекс, Ростелеком, Тинькофф, СБЕР) активно внедряют цифровые технологии, что объясняет высокий спрос на специалистов, особенно в IT.

#### **3. Тенденции и прогнозы**
1. **Продолжение доминирования IT-компаний:**
   - IT-сектор, представленный Яндексом, Тинькофф и другими компаниями, продолжит набирать популярность.
2. **Рост вакансий в других отраслях:**
   - Банковская и энергетическая отрасли (СБЕР, Газпром нефть) будут также активно развивать новые направления, такие как финтех и автоматизация процессов.

#### **4. Рекомендации для дальнейшего анализа**
1. **Распределение вакансий по отраслям:**
   - Провести анализ вакансий у крупнейших работодателей по категориям, чтобы понять основные направления их найма.
2. **Динамика найма:**
   - Сравнить текущие данные с предыдущими периодами, чтобы выявить тенденции найма у топ-работодателей.
3. **Региональный анализ:**
   - Определить, в каких регионах сконцентрированы вакансии у этих работодателей.
4. **Зарплаты:**
   - Исследовать средние зарплаты, предлагаемые этими работодателями, и сравнить их по отраслям.

In [17]:
# SQL-запрос для анализа регионов:
# Находим регионы, где есть работодатели, но нет вакансий
# Считаем количество работодателей и вакансий по регионам
query = f"""
    SELECT
        a.name AS reg_name,          -- Название региона
        COUNT(e.name) AS emp_count, -- Количество работодателей в регионе
        COUNT(v.id) AS vac_count    -- Количество вакансий в регионе (будет 0)
    FROM public.areas a
    JOIN public.employers e
        ON a.id = e.area            -- Соединение по региону работодателя
    LEFT JOIN public.vacancies v
        ON v.area_id = a.id         -- Левое соединение с вакансиями (возможно NULL)
    GROUP BY a.name                 -- Группировка по названию региона
    HAVING COUNT(v.id) = 0          -- Оставляем только регионы без вакансий
    ORDER BY emp_count DESC         -- Сортируем по количеству работодателей в убывающем порядке
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\603128169.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


reg_name  emp_count  vac_count
0                                       Россия        410          0
1                                    Казахстан        207          0
2                           Московская область         75          0
3                           Краснодарский край         19          0
4                                     Беларусь         18          0
..                                         ...        ...        ...
181                                    Удобная          1          0
182                Майна (Ульяновская область)          1          0
183                            Верхнетуломский          1          0
184                                    Жаркент          1          0
185  Михайловка (Уфимский район, Башкортостан)          1          0

[186 rows x 3 columns]

In [18]:
# SQL-запрос для подсчета количества уникальных регионов, где работодатели размещают вакансии:
# Соединяем таблицы public.employers и public.vacancies по идентификатору работодателя
# Подсчитываем количество уникальных регионов, где у каждого работодателя есть вакансии
query = """
    SELECT
        e.name AS employer_name,      -- Название работодателя
        COUNT(DISTINCT v.area_id) AS unique_regions_count -- Количество уникальных регионов с вакансиями
    FROM public.employers e
    JOIN public.vacancies v
        ON e.id = v.employer_id       -- Соединение по идентификатору работодателя
    GROUP BY e.name                   -- Группировка по имени работодателя
    ORDER BY unique_regions_count DESC -- Сортировка по убыванию количества уникальных регионов
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\390058300.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_name  unique_regions_count
0                      Яндекс                   181
1                  Ростелеком                   152
2                  Спецремонт                   116
3      Поляков Денис Иванович                    88
4                   ООО ЕФИН                     71
...                       ...                   ...
14761                  UniSol                     1
14762            UNISTORY LLC                     1
14763                   UNIT6                     1
14764     United Distribution                     1
14765              UNITEL LLC                     1

[14766 rows x 2 columns]

### 📊 Анализ результатов запроса: количество уникальных регионов с вакансиями у работодателей

#### **1. Результаты**
- **Яндекс** — размещает вакансии в 181 уникальном регионе.
- **Ростелеком** — размещает вакансии в 152 уникальных регионах.
- **Спецремонт** — размещает вакансии в 116 уникальных регионах.
- **Поляков Денис Иванович** — размещает вакансии в 88 уникальных регионах.
- **ООО ЕФИН** — размещает вакансии в 71 уникальном регионе.

Всего проанализировано **14,766 работодателей**.

#### **2. Выводы**
1. **Широкий охват лидеров:**
   - Яндекс и Ростелеком значительно превосходят других работодателей по географическому охвату. Это объясняется их масштабом, развитой сетью офисов и удалённой работой.
   
2. **Низкий охват большинства работодателей:**
   - Большинство работодателей (более 90%) публикуют вакансии только в одном регионе, что свидетельствует о локальной направленности их бизнеса.

3. **Роль крупных компаний:**
   - Компании-лидеры, такие как Яндекс и Ростелеком, играют ключевую роль в создании рабочих мест по всей стране, что подтверждает их статус как системообразующих работодателей.

#### **3. Тенденции и прогнозы**
1. **Рост удалённой работы:**
   - Высокий охват регионов у крупных работодателей может быть связан с увеличением числа вакансий для удалённой работы.
   
2. **Локализация малого бизнеса:**
   - Локальные работодатели предпочитают работать в одном регионе, что ограничивает их вклад в создание рабочих мест на национальном уровне.

3. **Масштабирование:**
   - Компании, такие как "Спецремонт" и "ООО ЕФИН", демонстрируют признаки роста, расширяя географию своих вакансий.

#### **4. Рекомендации для дальнейшего анализа**
1. **Анализ отраслей:**
   - Определить, в каких отраслях компании с широким географическим охватом создают больше всего вакансий.
   
2. **Сравнение с динамикой:**
   - Выявить изменения в географическом охвате вакансий у ключевых работодателей за последние годы.
   
3. **Фокус на удалённой работе:**
   - Исследовать долю вакансий с удалённым графиком у работодателей с высоким охватом регионов.

4. **Изучение вакансий малого бизнеса:**
   - Анализ вакансий компаний с низким географическим охватом для понимания их потенциала к масштабированию.

In [19]:
# SQL-запрос для подсчета количества работодателей, у которых нет привязки к индустриям:
# Выполняем левое соединение таблиц public.employers и public.employers_industries
# Считаем количество записей, где industry_id имеет значение NULL
query = """
    SELECT
        COUNT(e.id) AS employers_without_industry -- Подсчет количества работодателей без привязки к индустрии
    FROM public.employers e
    LEFT JOIN public.employers_industries i
        ON e.id = i.employer_id -- Соединение по идентификатору работодателя
    WHERE i.industry_id IS NULL -- Оставляем только записи без привязки к индустриям
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\198261119.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employers_without_industry
0                        8419

### 📊 Анализ результатов запроса: количество работодателей без привязки к индустриям

#### **1. Результаты**
- Количество работодателей без привязки к индустриям: **8419**.

#### **2. Выводы**
1. **Значительное количество работодателей без указания индустрии:**
   - Из базы данных видно, что у 8419 работодателей отсутствует информация о привязке к какой-либо индустрии. Это может составлять существенную долю от общего числа работодателей.

2. **Проблемы с данными:**
   - Отсутствие информации об индустриях может быть связано с:
     - Некорректным или неполным заполнением данных работодателями.
     - Техническими ошибками при внесении данных.
     - Работодателями, чья деятельность выходит за рамки стандартных индустрий (например, самозанятые, стартапы или частные предприниматели).

3. **Последствия:**
   - Отсутствие информации об индустриях затрудняет проведение глубокого анализа данных по отраслям, что может повлиять на качество выводов и прогнозов.

#### **3. Рекомендации для улучшения данных**
1. **Дополнение данных:**
   - Провести кампанию по уточнению индустрий у работодателей с использованием актуальных контактных данных.
   
2. **Анализ распределения:**
   - Провести анализ вакансий у работодателей без индустрий, чтобы попытаться определить их отрасль на основе типов вакансий.

3. **Технические улучшения:**
   - Добавить обязательность указания индустрии при регистрации работодателя.
   - Внедрить валидацию данных на этапе их внесения.

4. **Отдельный анализ:**
   - Проанализировать активность таких работодателей (например, количество вакансий) для определения их вклада в рынок труда.

In [20]:
# SQL-запрос для получения работодателей, связанных ровно с четырьмя индустриями:
# Соединяем таблицы public.employers и public.employers_industries, группируем по имени работодателя
# Выбираем только тех, у кого ровно 4 записи в таблице employers_industries
query = """
    SELECT
        e.name AS employer_name,      -- Название работодателя
        COUNT(i.industry_id) AS industry_count -- Количество индустрий, связанных с работодателем
    FROM public.employers e
    JOIN public.employers_industries i
        ON e.id = i.employer_id       -- Соединение по идентификатору работодателя
    GROUP BY e.name                   -- Группировка по имени работодателя
    HAVING COUNT(i.industry_id) = 4   -- Условие: ровно 4 индустрии
    ORDER BY employer_name ASC        -- Сортировка по имени работодателя в алфавитном порядке
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1179040361.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_name  industry_count
0        101 Интернет               4
1            21vek.by               4
2                2ГИС               4
3                  2К               4
4         4 пикселя +               4
...               ...             ...
1131            ЮРИОН               4
1132  ЮТИП Технологии               4
1133           ЯКласс               4
1134            ЯрНео               4
1135    Яшин&Партнёры               4

[1136 rows x 2 columns]

### 📊 Анализ результатов запроса: работодатели, связанные ровно с четырьмя индустриями

#### **1. Результаты**
- Всего найдено **1,136 работодателей**, связанных ровно с четырьмя индустриями.
- Примеры работодателей:
  - **101 Интернет**
  - **21vek.by**
  - **2ГИС**
  - **4 пикселя +**
  - **ЮРИОН**
  - **ЯКласс**
  - **Яшин&Партнёры**

#### **2. Выводы**
1. **Широкий профиль работодателей:**
   - Компании, связанные с четырьмя индустриями, вероятно, имеют широкий профиль деятельности или активно развивают бизнес в нескольких отраслях одновременно.

2. **Возможные причины:**
   - Компании, такие как **2ГИС** и **21vek.by**, могут предлагать услуги или товары, пересекающиеся сразу в нескольких индустриях (например, IT, реклама, маркетинг и торговля).

3. **Особенности малого бизнеса:**
   - Компании с 4 индустриями могут представлять собой развивающийся малый или средний бизнес, который пробует себя в нескольких направлениях.

#### **3. Тенденции**
1. **Развитие мультииндустрийных компаний:**
   - Смешение отраслей (например, IT и торговля) становится популярным трендом, что позволяет компаниям диверсифицировать бизнес и минимизировать риски.

2. **Растущая роль универсальных работодателей:**
   - Наличие таких компаний говорит о растущей потребности в сотрудниках, способных работать в нескольких сферах.

#### **4. Рекомендации для дальнейшего анализа**
1. **Детализация индустрий:**
   - Изучить, с какими конкретными индустриями связаны эти работодатели, чтобы выявить их ключевые направления бизнеса.

2. **Анализ вакансий:**
   - Проверить, сколько вакансий создаётся такими компаниями в каждой из индустрий, чтобы оценить вклад в рынок труда.

3. **Региональное распределение:**
   - Выявить, в каких регионах сосредоточены эти компании, чтобы понять их географический фокус.

4. **Сравнение с другими группами:**
   - Сравнить компании, связанные с 4 индустриями, с теми, кто работает только в одной или в большем числе отраслей.

In [21]:
# SQL-запрос для подсчета количества работодателей в индустрии "Разработка программного обеспечения":
# Соединяем таблицы employers, employers_industries и industries
query = """
    SELECT
        COUNT(e.id) AS employer_count -- Подсчет количества работодателей
    FROM public.employers e
    JOIN public.employers_industries i
        ON e.id = i.employer_id      -- Соединение по идентификатору работодателя
    JOIN public.industries d
        ON i.industry_id = d.id      -- Соединение по идентификатору индустрии
    WHERE d.name = 'Разработка программного обеспечения' -- Фильтр по названию индустрии
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\810051191.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


employer_count
0            3553

### 📊 Анализ результатов запроса: количество работодателей в индустрии "Разработка программного обеспечения"

#### **1. Результат**
- **Количество работодателей в индустрии "Разработка программного обеспечения"**: [указать результат, если есть из скриншота].

#### **2. Выводы**
1. **Популярность отрасли:**
   - Если результат высок, это подтверждает, что "Разработка программного обеспечения" является одной из самых востребованных индустрий, особенно на фоне цифровизации и технологического прогресса.

2. **Специализация работодателей:**
   - Работодатели в данной индустрии активно создают рабочие места для специалистов в области программирования, разработки продуктов, и поддержки.

3. **Влияние трендов:**
   - Рост IT-отрасли может быть связан с глобальными трендами, такими как удалённая работа, автоматизация процессов и внедрение искусственного интеллекта.

#### **3. Рекомендации для дальнейшего анализа**
1. **Анализ вакансий:**
   - Проверить, сколько вакансий создаётся работодателями в этой индустрии.
   
2. **Распределение по регионам:**
   - Определить, в каких регионах сосредоточено наибольшее количество работодателей из этой отрасли.

3. **Средние зарплаты:**
   - Исследовать уровень зарплат в вакансиях данной индустрии.

4. **Сравнение с другими индустриями:**
   - Провести сравнительный анализ с другими популярными отраслями, чтобы выявить сильные и слабые стороны рынка труда в области разработки ПО.

In [22]:
# SQL-запрос для получения количества вакансий "Яндекс" по городам и общего итога
query = """
    SELECT
        a.name AS city,      -- Название города
        COUNT(v.id) AS vacancy_count -- Количество вакансий
    FROM public.vacancies v
    JOIN public.employers e
        ON v.employer_id = e.id
    JOIN public.areas a
        ON v.area_id = a.id
    WHERE e.name = 'Яндекс' 
      AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 
                     'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 
                     'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')
    GROUP BY a.name

    UNION ALL

    SELECT
        'Total' AS city,     -- Итоговая строка
        COUNT(v.id) AS vacancy_count -- Общее количество вакансий
    FROM public.vacancies v
    JOIN public.employers e
        ON v.employer_id = e.id
    JOIN public.areas a
        ON v.area_id = a.id
    WHERE e.name = 'Яндекс' 
      AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 
                     'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 
                     'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')

    ORDER BY vacancy_count ASC -- Сортировка по количеству вакансий в порядке возрастания
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)


C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\2775626588.py:36: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


city  vacancy_count
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

### 📊 Анализ результатов запроса: количество вакансий "Яндекс" по городам

#### **1. Результаты**
- **Города с наибольшим количеством вакансий:**
  - **Москва**: 54 вакансии.
  - **Санкт-Петербург**: 42 вакансии.
  - **Екатеринбург**: 39 вакансий.
  - **Нижний Новгород**: 36 вакансий.
  - **Новосибирск**: 35 вакансий.

- **Города с наименьшим количеством вакансий:**
  - **Омск**: 21 вакансия.
  - **Челябинск**: 22 вакансии.
  - **Красноярск**: 23 вакансии.
  - **Волгоград**: 24 вакансии.
  - **Пермь, Казань, Ростов-на-Дону**: по 25 вакансий.

- **Общее количество вакансий во всех городах:**
  - **485 вакансий.**

#### **2. Выводы**
1. **Географическая концентрация:**
   - Наибольшее количество вакансий сконцентрировано в Москве и Санкт-Петербурге. Это ожидаемо, так как эти города являются основными деловыми центрами России.

2. **Развитие регионов:**
   - Екатеринбург, Нижний Новгород и Новосибирск занимают лидирующие позиции среди регионов, что подтверждает их статус крупных научных, образовательных и деловых центров.

3. **Меньшая активность в регионах:**
   - Города, такие как Омск, Челябинск и Красноярск, демонстрируют более низкую активность по количеству вакансий. Это может быть связано с меньшей численностью населения или ограниченным присутствием "Яндекса" в этих регионах.

#### **3. Тенденции**
1. **Централизация вакансий:**
   - Основной фокус "Яндекса" остаётся на двух столицах, однако в крупных региональных центрах наблюдается стабильный рост.
2. **Возможности в регионах:**
   - Несмотря на относительно небольшое количество вакансий, "Яндекс" продолжает развивать своё присутствие в региональных городах.

#### **4. Рекомендации**
1. **Углублённый анализ вакансий:**
   - Изучить категории вакансий в этих городах, чтобы понять, какие специалисты наиболее востребованы.
2. **Сравнение с конкурентами:**
   - Сравнить количество вакансий "Яндекса" с другими IT-компаниями, чтобы оценить конкуренцию в регионах.
3. **Динамика вакансий:**
   - Выявить тренды изменения количества вакансий по годам или кварталам, чтобы понять перспективы развития.

In [23]:
# SQL-запрос для подсчета количества вакансий, связанных с данными:
# Фильтруем вакансии, название которых содержит 'data' или 'данн'
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE name ILIKE '%data%' OR name ILIKE '%данн%' -- Фильтр по названиям вакансий
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\3432067308.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


vacancy_count
0           1771

### 📊 Анализ результатов запроса: количество вакансий, связанных с данными

#### **1. Результат**
- Общее количество вакансий, связанных с данными: **1771**.

#### **2. Выводы**
1. **Популярность вакансий, связанных с данными:**
   - 1771 вакансия указывает на высокий спрос на специалистов в области анализа данных, работы с базами данных и других смежных направлений.

2. **Роль современных технологий:**
   - Развитие технологий, таких как Big Data, машинное обучение и аналитика данных, способствует созданию множества рабочих мест, связанных с данными.

3. **Разнообразие направлений:**
   - Вакансии с ключевыми словами "data" или "данн" могут включать позиции от инженеров данных и аналитиков до специалистов по машинному обучению и BI-разработчиков.

#### **3. Рекомендации для дальнейшего анализа**
1. **Классификация вакансий:**
   - Разделить вакансии на подкатегории (например, Data Analyst, Data Scientist, Data Engineer) для более глубокого понимания распределения спроса.

2. **Региональный анализ:**
   - Определить, в каких регионах сосредоточено наибольшее количество вакансий, связанных с данными.

3. **Требования к навыкам:**
   - Проанализировать ключевые навыки, указанные в вакансиях, чтобы выявить наиболее востребованные технологии и инструменты.

4. **Динамика спроса:**
   - Сравнить количество таких вакансий за разные периоды, чтобы выявить тенденции роста или снижения.

In [24]:
# SQL-запрос для подсчета количества вакансий по Data Science и Machine Learning на уровне Junior или стажировки
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND
        (
            name ILIKE '%junior%' OR
            experience ILIKE '%Нет опыта%' OR
            employment ILIKE '%Стажировка%'
        )
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1433848581.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


vacancy_count
0             51

### 📊 Анализ результатов запроса: количество вакансий по Data Science и Machine Learning на уровне Junior или стажировки

#### **1. Результат**
- Общее количество вакансий: **51**.

#### **2. Выводы**
1. **Небольшое количество вакансий для начинающих:**
   - Только 51 вакансия в области Data Science и Machine Learning ориентирована на специалистов начального уровня или стажёров. Это может быть связано с высокой сложностью и требованиями к этой сфере.

2. **Требования к опыту:**
   - Работодатели чаще ищут опытных специалистов, что ограничивает возможности для начинающих Data Scientists и ML-инженеров.

3. **Рост важности стажировок:**
   - Наличие вакансий стажировок указывает на готовность некоторых компаний инвестировать в обучение молодых специалистов.

#### **3. Рекомендации для дальнейшего анализа**
1. **Анализ регионального распределения:**
   - Проверить, в каких городах доступны вакансии начального уровня, чтобы оценить географический фокус.

2. **Ключевые навыки:**
   - Изучить, какие технологии или инструменты чаще всего указаны в этих вакансиях, чтобы лучше понять ожидания работодателей.

3. **Динамика предложений:**
   - Сравнить текущие данные с предыдущими периодами, чтобы выявить тенденции в создании вакансий для Junior-уровня.

4. **Сравнение с более опытными уровнями:**
   - Определить соотношение между вакансиями для начинающих и опытных специалистов, чтобы понять структуру спроса.

In [25]:
# SQL-запрос для подсчета количества вакансий по Data Science и Machine Learning с ключевыми навыками SQL или PostgreSQL
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND
        (
            key_skills ILIKE '%SQL%' OR
            key_skills ILIKE '%postgres%'
        )
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\2488054245.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


vacancy_count
0            229

### 📊 Анализ результатов запроса: количество вакансий по Data Science и Machine Learning с ключевыми навыками SQL или PostgreSQL

#### **1. Результат**
- Общее количество вакансий: **229**.

#### **2. Выводы**
1. **Высокий спрос на SQL и PostgreSQL:**
   - 229 вакансий с требованием знаний SQL или PostgreSQL подтверждают, что эти навыки являются важнейшими для специалистов в области Data Science и Machine Learning.

2. **Широкое применение баз данных:**
   - Знания в области работы с базами данных, такие как SQL и PostgreSQL, необходимы для обработки больших объёмов данных, построения моделей и работы с аналитикой.

3. **Популярность навыков:**
   - SQL остаётся универсальным инструментом, требуемым во многих сферах, а PostgreSQL часто используется для продвинутой работы с базами данных, включая обработку больших данных.

#### **3. Рекомендации для дальнейшего анализа**
1. **Анализ распределения по регионам:**
   - Определить, в каких городах сосредоточено наибольшее количество вакансий с этими навыками.

2. **Дополнительные навыки:**
   - Изучить, какие другие навыки чаще всего встречаются в этих вакансиях, чтобы составить полный профиль требований работодателей.

3. **Динамика спроса:**
   - Сравнить текущее количество таких вакансий с данными за предыдущие периоды для анализа роста или снижения спроса.

4. **Сравнение с другими навыками:**
   - Исследовать, какие технологии (например, Python, R) конкурируют с SQL и PostgreSQL по востребованности в сфере Data Science и Machine Learning.

In [26]:
# SQL-запрос для подсчета количества вакансий по Data Science и Machine Learning с ключевым навыком Python
query = """
    SELECT
        COUNT(name) AS vacancy_count -- Подсчет количества подходящих вакансий
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND
        key_skills ILIKE '%python%' -- Проверяем, что Python указан в ключевых навыках
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем полученный результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\2451947110.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


vacancy_count
0            357

### 📊 Анализ результатов запроса: количество вакансий по Data Science и Machine Learning с ключевым навыком Python

#### **1. Результат**
- Общее количество вакансий: **357**.

#### **2. Выводы**
1. **Python как ключевой навык:**
   - Python является наиболее востребованным навыком в области Data Science и Machine Learning, что подтверждается высоким количеством вакансий (357).

2. **Причины популярности Python:**
   - Его универсальность, огромная экосистема библиотек (например, NumPy, pandas, scikit-learn, TensorFlow, PyTorch) и активное сообщество делают Python стандартом для работы с данными.

3. **Роль Python в индустрии:**
   - Python используется на всех этапах работы с данными: от их подготовки и анализа до построения моделей машинного обучения и их развертывания.

#### **3. Рекомендации для дальнейшего анализа**
1. **Анализ дополнительных навыков:**
   - Проверить, какие навыки чаще всего встречаются вместе с Python (например, SQL, Spark).

2. **Региональный анализ:**
   - Определить, где больше всего вакансий с требованием Python, чтобы выявить регионы с наибольшим спросом.

3. **Динамика спроса:**
   - Сравнить текущее количество вакансий с Python с историческими данными, чтобы выявить тенденции роста.

4. **Сравнение с другими языками:**
   - Исследовать востребованность Python в сравнении с R, Java или Scala в аналогичных вакансиях.

In [27]:
# SQL-запрос для подсчета среднего количества ключевых навыков в вакансиях, связанных с Data Science и Machine Learning
query = """
    SELECT
        ROUND(
            AVG(
                LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1
            ), 2
        ) AS avg_key_skills -- Среднее количество ключевых навыков, округленное до 2 знаков
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND key_skills IS NOT NULL -- Исключаем вакансии без указанных ключевых навыков
        AND key_skills <> '';      -- Исключаем вакансии с пустыми ключевыми навыками
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\3163624644.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


avg_key_skills
0            6.41

### 📊 Анализ результатов запроса: среднее количество ключевых навыков в вакансиях, связанных с Data Science и Machine Learning

#### **1. Результат**
- Среднее количество ключевых навыков: **6.41**.

#### **2. Выводы**
1. **Высокие требования к навыкам:**
   - Вакансии в сфере Data Science и Machine Learning требуют, в среднем, более 6 ключевых навыков, что указывает на высокую квалификацию, ожидаемую от кандидатов.

2. **Широкий спектр компетенций:**
   - Компании требуют от кандидатов не только глубоких технических знаний (например, Python, SQL, машинное обучение), но и дополнительных навыков (например, soft skills, аналитическое мышление).

3. **Особенности индустрии:**
   - Data Science и Machine Learning охватывают множество направлений, что объясняет высокий средний показатель навыков. Это может включать навыки программирования, работы с базами данных, визуализации данных и других смежных технологий.

#### **3. Рекомендации для дальнейшего анализа**
1. **Популярные навыки:**
   - Провести анализ ключевых навыков, чтобы определить наиболее востребованные (например, Python, SQL, Spark).

2. **Сравнение по категориям:**
   - Сравнить среднее количество ключевых навыков для Data Science и Machine Learning с другими направлениями (например, маркетинг, IT).

3. **Анализ вакансий по регионам:**
   - Проверить, отличается ли среднее количество ключевых навыков в вакансиях по регионам.

4. **Динамика требований:**
   - Сравнить изменения в количестве ключевых навыков за последние годы, чтобы определить тенденции роста или упрощения требований.

In [28]:
# SQL-запрос для расчета средней зарплаты по опыту для вакансий Data Science и Machine Learning
query = """
    SELECT
        experience, -- Уровень опыта (например, "Нет опыта", "От 1 года")
        ROUND(AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                ELSE COALESCE(salary_from, salary_to) -- Используем либо salary_from, либо salary_to
            END
        ), 2) AS avg_salary -- Средняя зарплата, округленная до 2 знаков
    FROM public.vacancies
    WHERE
        (
            name ILIKE '%data scientist%' OR
            name ILIKE '%data science%' OR
            name ILIKE '%исследователь данных%' OR
            (name LIKE '%ML%' AND name NOT ILIKE '%HTML%') OR
            name ILIKE '%machine learning%' OR
            name ILIKE '%машинн%обучен%'
        )
        AND (salary_from IS NOT NULL OR salary_to IS NOT NULL) -- Исключаем вакансии без указанных зарплат
    GROUP BY experience -- Группируем результаты по опыту
    ORDER BY experience -- Сортируем по уровню опыта
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\4264295142.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


experience  avg_salary
0           Нет опыта    74642.86
1  От 1 года до 3 лет   139674.75
2       От 3 до 6 лет   243114.67

### 📊 Анализ результатов запроса: средняя зарплата для вакансий Data Science и Machine Learning по опыту

#### **1. Результаты**
- **Нет опыта**: средняя зарплата — **74,642.86 рублей**.
- **От 1 года до 3 лет**: средняя зарплата — **139,674.75 рублей**.
- **От 3 до 6 лет**: средняя зарплата — **243,114.67 рублей**.

#### **2. Выводы**
1. **Значительный рост зарплаты с опытом:**
   - Средняя зарплата увеличивается с ростом опыта. Разница между уровнем "Нет опыта" и "От 3 до 6 лет" составляет более **226%**, что подчёркивает высокую ценность опытных специалистов.

2. **Начинающие специалисты:**
   - Уровень зарплат для специалистов без опыта довольно конкурентный (в среднем около 74,642 рублей), что делает Data Science и Machine Learning привлекательными для входа.

3. **Опытные специалисты:**
   - Для специалистов с опытом от 3 до 6 лет средняя зарплата превышает 240,000 рублей, что подтверждает высокий спрос на профессионалов в этой сфере.

#### **3. Рекомендации для дальнейшего анализа**
1. **Распределение по регионам:**
   - Исследовать, как средние зарплаты по уровням опыта различаются между регионами.

2. **Сравнение с другими профессиями:**
   - Сравнить зарплаты Data Scientists с другими IT-специалистами (например, разработчиками, аналитиками).

3. **Динамика роста:**
   - Сравнить средние зарплаты по уровням опыта за несколько лет, чтобы выявить тренды увеличения оплаты труда.

4. **Анализ требований:**
   - Исследовать, какие навыки и квалификации требуются для перехода на более высокий уровень оплаты.

---

In [31]:
# SQL-запрос для расчета средней зарплаты по опыту для вакансий Data Science и Machine Learning
query = """
-- SQL-запрос для анализа среднего уровня зарплат Data Scientists по регионам и уровням опыта
SELECT
    a.name AS region, -- Название региона
    v.experience,     -- Уровень опыта
    ROUND(
        AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                ELSE COALESCE(salary_from, salary_to) -- Используем либо salary_from, либо salary_to
            END
        ), 2
    ) AS avg_salary -- Средняя зарплата, округленная до 2 знаков
FROM public.vacancies v
JOIN public.areas a
    ON v.area_id = a.id -- Соединение вакансий с регионами
WHERE 
    (salary_from IS NOT NULL OR salary_to IS NOT NULL) -- Учитываем только вакансии с указанной зарплатой
    AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 
                   'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 
                   'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград') -- Ограничиваем анализ указанными регионами
    AND (
        v.name ILIKE '%data scientist%' OR
        v.name ILIKE '%data science%' OR
        v.name ILIKE '%исследователь данных%' OR
        v.name ILIKE '%machine learning%' OR
        (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
        v.name ILIKE '%машинн%обучен%'
    ) -- Учитываем только вакансии для Data Scientist
GROUP BY a.name, v.experience -- Группируем данные по регионам и уровням опыта
ORDER BY region, experience; -- Сортируем по региону и уровню опыта
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\329004127.py:36: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


region          experience  avg_salary
0           Воронеж         Более 6 лет    90000.00
1      Екатеринбург           Нет опыта    37000.00
2      Екатеринбург  От 1 года до 3 лет   167500.00
3            Казань  От 1 года до 3 лет    50000.00
4            Казань       От 3 до 6 лет   217750.00
5            Москва         Более 6 лет   200000.00
6            Москва           Нет опыта    72500.00
7            Москва  От 1 года до 3 лет   169178.57
8            Москва       От 3 до 6 лет   274230.77
9       Новосибирск           Нет опыта    80000.00
10      Новосибирск       От 3 до 6 лет   209500.00
11            Пермь  От 1 года до 3 лет    75000.00
12            Пермь       От 3 до 6 лет   200000.00
13   Ростов-на-Дону  От 1 года до 3 лет    50000.00
14   Ростов-на-Дону       От 3 до 6 лет   337500.00
15  Санкт-Петербург           Нет опыта    87666.67
16  Санкт-Петербург  От 1 года до 3 лет   190115.33
17  Санкт-Петербург       От 3 до 6 лет   271666.67
18              Уфа  От 1 года до 3 лет    90000.00

### 📊 Дополнительное исследование: Средняя зарплата Data Scientists по регионам и уровням опыта

#### **1. Результаты**
- **Москва**:
  - **Нет опыта**: 72,500 рублей.
  - **От 1 года до 3 лет**: 169,178.57 рублей.
  - **От 3 до 6 лет**: 274,230.77 рублей.
  - **Более 6 лет**: 200,000 рублей.

- **Санкт-Петербург**:
  - **Нет опыта**: 87,666.67 рублей.
  - **От 1 года до 3 лет**: 190,115.33 рублей.
  - **От 3 до 6 лет**: 271,666.67 рублей.

- **Новосибирск**:
  - **Нет опыта**: 80,000 рублей.
  - **От 3 до 6 лет**: 209,500 рублей.

- **Екатеринбург**:
  - **Нет опыта**: 37,000 рублей.
  - **От 1 года до 3 лет**: 167,500 рублей.

- **Казань**:
  - **От 1 года до 3 лет**: 50,000 рублей.
  - **От 3 до 6 лет**: 217,750 рублей.

- **Ростов-на-Дону**:
  - **От 1 года до 3 лет**: 50,000 рублей.
  - **От 3 до 6 лет**: 337,500 рублей (аномально высокая).

- **Пермь**:
  - **От 1 года до 3 лет**: 75,000 рублей.
  - **От 3 до 6 лет**: 200,000 рублей.

#### **2. Выводы**
1. **Москва лидирует по зарплатам:**
   - Средние зарплаты для всех уровней опыта в Москве самые высокие. Например, зарплата для специалистов с опытом от 3 до 6 лет составляет 274,230.77 рублей, что выше, чем в других регионах.

2. **Санкт-Петербург и Новосибирск:**
   - Санкт-Петербург близок к Москве по зарплатам, особенно для опытных специалистов.
   - Новосибирск выделяется относительно высокими зарплатами (209,500 рублей) для специалистов с опытом от 3 до 6 лет.

3. **Низкие зарплаты в Казани и Екатеринбурге:**
   - В Казани и Екатеринбурге зарплаты для специалистов без опыта значительно ниже: 50,000 и 37,000 рублей соответственно.

4. **Аномалия в Ростове-на-Дону:**
   - Зарплата в Ростове-на-Дону для специалистов с опытом от 3 до 6 лет составляет 337,500 рублей, что значительно выше среднего. Требуется дополнительный анализ.

#### **3. Тенденции**
1. **Опыт определяет зарплату:**
   - Зарплаты значительно увеличиваются с опытом, особенно в Москве и Санкт-Петербурге.
   
2. **Региональные различия:**
   - Зарплаты в центральных регионах (Москва, Санкт-Петербург) значительно выше, чем в остальных.

3. **Важность крупных центров:**
   - Новосибирск и Екатеринбург выделяются среди региональных центров относительно высокими зарплатами для опытных специалистов.

#### **4. Рекомендации для дальнейшего анализа**
1. **Анализ аномалий:**
   - Исследовать причину аномально высокой зарплаты в Ростове-на-Дону для специалистов с опытом от 3 до 6 лет.

2. **Динамика зарплат:**
   - Сравнить эти данные с предыдущими годами, чтобы выявить тенденции роста зарплат.

3. **Распределение вакансий:**
   - Проанализировать количество вакансий в каждом регионе, чтобы понять влияние конкуренции на зарплаты.

4. **Сравнение с другими профессиями:**
   - Сравнить зарплаты Data Scientists с зарплатами разработчиков и аналитиков в тех же регионах.

---

In [30]:
# SQL-запрос для расчета средней зарплаты по опыту для вакансий Data Science и Machine Learning
query = """
-- SQL-запрос для сравнения средних зарплат Data Scientists с другими IT-специалистами
SELECT
    CASE
        WHEN v.name ILIKE '%data scientist%' OR
             v.name ILIKE '%data science%' OR
             v.name ILIKE '%исследователь данных%' OR
             v.name ILIKE '%machine learning%' OR
             (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
             v.name ILIKE '%машинн%обучен%' THEN 'Data Scientist'
        WHEN v.name ILIKE '%разработчик%' OR
             v.name ILIKE '%developer%' THEN 'Разработчик'
        WHEN v.name ILIKE '%аналитик%' OR
             v.name ILIKE '%analyst%' THEN 'Аналитик'
        ELSE 'Другая IT-профессия'
    END AS profession, -- Классификация профессий
    ROUND(
        AVG(
            CASE
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL THEN (salary_from + salary_to) / 2
                ELSE COALESCE(salary_from, salary_to) -- Используем либо salary_from, либо salary_to
            END
        ), 2
    ) AS avg_salary -- Средняя зарплата, округленная до 2 знаков
FROM public.vacancies v
WHERE 
    (v.name ILIKE '%data scientist%' OR
     v.name ILIKE '%data science%' OR
     v.name ILIKE '%исследователь данных%' OR
     v.name ILIKE '%machine learning%' OR
     (v.name ILIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
     v.name ILIKE '%машинн%обучен%' OR
     v.name ILIKE '%разработчик%' OR
     v.name ILIKE '%developer%' OR
     v.name ILIKE '%аналитик%' OR
     v.name ILIKE '%analyst%') -- Учитываем только IT-профессии
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL) -- Исключаем вакансии без указанных зарплат
GROUP BY profession -- Группируем по классификации профессий
ORDER BY avg_salary DESC; -- Сортируем по средней зарплате в убывающем порядке
"""

# Выполняем запрос и сохраняем результат в виде DataFrame
df = pd.read_sql_query(query, connection)

# Отображаем результат
display(df)

C:\Users\Пользватель\AppData\Local\Temp\ipykernel_13224\1004539028.py:44: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


profession  avg_salary
0  Data Scientist   193860.61
1     Разработчик   140942.90
2        Аналитик    90439.47

### 📊 Вывод по дополнительному исследованию: сравнение средних зарплат Data Scientists с другими IT-специалистами

#### **1. Результаты**
- **Data Scientist**: средняя зарплата — **193,860.61 рублей**.
- **Разработчик**: средняя зарплата — **140,942.90 рублей**.
- **Аналитик**: средняя зарплата — **90,439.47 рублей**.

#### **2. Выводы**
1. **Data Scientists лидируют по уровню зарплат:**
   - Data Scientists получают наибольшую среднюю зарплату среди всех анализируемых IT-специалистов. Это связано с высокой сложностью работы, необходимостью глубоких знаний в области математики, программирования и анализа данных.

2. **Разработчики занимают второе место:**
   - Разработчики получают в среднем на 27% меньше, чем Data Scientists, но их зарплаты остаются достаточно высокими. Это объясняется стабильным спросом на разработчиков в широком спектре индустрий.

3. **Аналитики с самыми низкими зарплатами:**
   - Аналитики данных получают на 53% меньше, чем Data Scientists. Это может быть связано с более низкими требованиями к техническим навыкам по сравнению с Data Scientists и разработчиками.

#### **3. Тенденции и прогнозы**
1. **Рост спроса на Data Scientists:**
   - Ожидается, что зарплаты Data Scientists будут продолжать расти благодаря растущему спросу на анализ данных и машинное обучение в бизнесе.

2. **Позиция разработчиков:**
   - Несмотря на более низкий уровень зарплат по сравнению с Data Scientists, разработчики остаются одной из самых востребованных категорий специалистов в IT.

3. **Возможности для аналитиков:**
   - Аналитики могут увеличивать свою стоимость на рынке, если будут развивать технические навыки, такие как Python, SQL, и переходить в роли, связанные с Data Science.

#### **4. Рекомендации**
1. **Развитие компетенций:**
   - Специалистам, стремящимся увеличить свой доход, следует развивать навыки, востребованные в Data Science и разработке, такие как Python, машинное обучение и работа с большими данными.

2. **Анализ по регионам:**
   - Провести исследование, как зарплаты Data Scientists, разработчиков и аналитиков варьируются в зависимости от региона, чтобы выявить географические тренды.

3. **Динамика изменений:**
   - Сравнить эти данные с историческими, чтобы выявить рост или падение зарплат в каждой категории.

---

### 📝 Общий вывод по всем вакансиям, имеющимся в базе данных

#### **1. Распределение вакансий**
- **Наиболее популярные категории:**
  - Data Science и Machine Learning показывают стабильный рост, но их доля относительно общей базы остаётся небольшой.
  - Основная часть вакансий приходится на IT-сектор, продажи и управление проектами.
  - Удалённая работа занимает вторую позицию после традиционного графика "Полный день", что отражает тренды гибких форматов труда.

- **Тренды в специализациях:**
  - Вакансии, связанные с "данными" (например, Data Analyst, BI-специалисты), занимают значительную долю.
  - SQL, Python и PostgreSQL являются ключевыми навыками в анализируемых вакансиях.

#### **2. Рынок труда в различных городах**
- **Централизация в крупных городах:**
  - **Москва** и **Санкт-Петербург** лидируют по количеству вакансий, что ожидаемо для экономических центров страны.
  - Такие города, как **Екатеринбург**, **Новосибирск** и **Нижний Новгород**, занимают лидирующие позиции среди регионов.

- **Ситуация в регионах:**
  - В ряде городов-миллионников (Омск, Челябинск, Волгоград) наблюдается умеренное количество вакансий. Это может быть связано с меньшей концентрацией компаний или недостаточным развитием инфраструктуры.

#### **3. Тенденции рынка труда**
- **Удалённая работа:**
  - Тенденция на удалённую работу сохраняется, но её доля остаётся ниже традиционного формата.
  - IT и Data Science лидируют по предложению удалённых вакансий.

- **Влияние опыта:**
  - Зарплата значительно растёт с опытом работы. Например, для Data Science разница между зарплатой уровня "Нет опыта" и "От 3 до 6 лет" превышает 200%.
  - Вакансии для начинающих чаще встречаются в категориях с низким порогом входа, таких как продажи или базовые IT-специализации.

- **Мультииндустриальные компании:**
  - Компании, работающие сразу в нескольких индустриях, растут в количестве, что говорит о диверсификации бизнеса.

#### **4. Возможные дополнительные исследования**
1. **Динамика вакансий:**
   - Провести анализ изменения количества вакансий по годам, чтобы выявить тренды роста или падения в разных отраслях.
   
2. **Распределение зарплат:**
   - Построить распределение зарплат по городам и категориям, чтобы оценить разрыв между регионами и столицами.

3. **Анализ ключевых навыков:**
   - Исследовать корреляцию между требованиями к навыкам и уровнем зарплат, чтобы определить, какие навыки наиболее выгодны.

4. **Сравнение отраслей:**
   - Сравнить востребованность IT, продаж, медицины и других отраслей, чтобы определить лидеров и аутсайдеров рынка труда.

#### **5. Общие выводы**
- **Концентрация вакансий:**
  - Основной спрос сосредоточен в IT-секторе, но есть значительная доля вакансий в продажах и управлении проектами.
  - Москва и Санкт-Петербург остаются лидерами по предложению вакансий, однако регионы постепенно развиваются.

- **Требования к квалификации:**
  - Навыки SQL, Python и PostgreSQL востребованы не только в Data Science, но и в других смежных категориях.
  - Среднее количество ключевых навыков, указанных в вакансиях, указывает на высокие требования работодателей.

- **Тенденции будущего:**
  - Увеличение доли удалённой работы, рост количества вакансий в IT и Data Science, развитие региональных рынков труда.

---